File: src/leetcode/121.best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.rs
use std::cmp;
struct Solution;
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        if prices.is_empty() {
            return 0;
        }

        let mut max_profit = 0;
        let mut min_price = prices[0];

        for &price in prices.iter() {
            min_price = cmp::min(price, min_price);
            max_profit = cmp::max(price - min_price, max_profit)
        }
        max_profit
    }
}
fn main() {
    let prices = vec![7, 1, 5, 3, 6, 4];
    Solution::max_profit(prices);
}

#[cfg(test)]
mod tests {
    use super::Solution;

    // Write your tests here
    #[test]
    fn test_case_1() {
        let prices = vec![7, 1, 5, 3, 6, 4];
        let result = 5;
        assert_eq!(result, Solution::max_profit(prices));
    }

    #[test]
    fn test_case_2() {
        let prices = vec![7, 6, 4, 3, 1];
        let result = 0;
        assert_eq!(result, Solution::max_profit(prices));
    }
    #[test]
    fn test_case_3_empty_prices() {
        let prices = vec![];
        let result = 0;
        assert_eq!(result, Solution::max_profit(prices));
    }
}

File: src/leetcode/121.best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.py
import unittest
from typing import *


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0

        max_profit: int = 0
        min_price = prices[0]

        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)

        return max_profit


class TestBestTimeToBuyAndSellStock(unittest.TestCase):
    def test_case_1(self):
        prices = [7, 1, 5, 3, 6, 4]
        result = 5
        self.assertEqual(result, Solution().maxProfit(prices))

    def test_case_2(self):
        prices = [7, 6, 4, 3, 1]
        result = 0
        self.assertEqual(result, Solution().maxProfit(prices))

    def test_case_3_empty_prices(self):
        prices = []
        result = 0
        self.assertEqual(result, Solution().maxProfit(prices))


if __name__ == "__main__":
    unittest.main()
